File: Octave/BuildCosMatrix.m
function [cosMatrix xVector]= BuildCosMatrix(xCoords, yCoords, zCoords, ElementData, vectorLevel)
            xDiff=xCoords(end)-xCoords(1);
            yDiff=yCoords(end)-yCoords(1);
            zDiff=zCoords(end)-zCoords(1);
            DiffVector=[xDiff;yDiff;zDiff];
           %Old implementation retained for understading new implementation below. 
           % if (vectorLevel==1)
           %     ElementLength=(xDiff);
           %     cos_xyz=1;
           % elseif (vectorLevel==2)
           %     ElementLength=sqrt(xDiff^2+yDiff^2);
           %     cos_xyz=[xDiff,yDiff]*(1/ElementLength);
           % elseif (vectorLevel==3)
           %     ElementLength=sqrt(xDiff^2+yDiff^2+zDiff^2);
           %     cos_xyz=[xDiff,yDiff,zDiff]*(1/ElementLength);
           % end

            SquareOfDiffVector=DiffVector.^2;
            SumOfSquare=sum(SquareOfDiffVector(1:vectorLevel),1); %vector Squared only upto the specified Vector Level
            ElementLength=sqrt(SumOfSquare);
            cos_xyz=DiffVector(1:vectorLevel)'./ElementLength; %vector Saved only upto the specified Vector Level
            cosMatrix=zeros(ElementData.NumOfElementNodes,vectorLevel*ElementData.NumOfElementNodes);
            xVector=zeros(vectorLevel*ElementData.NumOfElementNodes,1);
            CoordsTempStore=[xCoords, yCoords, zCoords]; %Temporary storage of coordiantes.
            for Row=1:ElementData.NumOfElementNodes;
                Coords=CoordsTempStore(Row,1:vectorLevel)'; %vector Saved only upto the specified Vector Level
                cosMatrix(Row,(vectorLevel*Row-(vectorLevel-1)):(vectorLevel*Row))=cos_xyz;  %A First Course in the Finite Element Method - Daryl L. Logan eqn 3.7.8; [... 0 0 Cx Cy Cz 0 0 ...]
                xVector((vectorLevel*Row-(vectorLevel-1)):(vectorLevel*Row),1)= Coords;
            end
end

File: Cpp/FEMtools/BuildCosMatrix.cpp
// Automatically translated using m2cpp 2.0 on 2018-12-13 19:09:24

#ifndef BUILDCOSMATRIX_M_HPP
#define BUILDCOSMATRIX_M_HPP


#include <armadillo>
#include "libGmshReader.h"
using namespace arma ;


void BuildCosMatrix(mat xCoords, mat yCoords, mat zCoords, libGmshReader::ElementData ElementData1, int i, int vectorLevel, mat& cosMatrix, mat& xVector)
{
    mat Coords, CoordsTempStore, DiffVector, ElementLength, SquareOfDiffVector, SumOfSquare, cos_xyz, xDiff, yDiff, zDiff ;
    int Row;
    xDiff = xCoords(xCoords.n_rows-1,0)-xCoords(0,0) ;
    yDiff = yCoords(yCoords.n_rows-1,0)-yCoords(0,0) ;
    zDiff = zCoords(zCoords.n_rows-1,0)-zCoords(0,0) ;
    DiffVector=join_cols( xDiff,join_rows(yDiff,zDiff));
    SquareOfDiffVector = arma::square(DiffVector) ;
    SumOfSquare = sum(SquareOfDiffVector(span(0,vectorLevel), 0)) ;
    ElementLength = sqrt(SumOfSquare) ;
    cos_xyz = trans(DiffVector(0,span(0, vectorLevel)))/ElementLength ;
    cosMatrix = zeros<mat>(ElementData1.NumOfElementNodes[i], vectorLevel*ElementData1.NumOfElementNodes[i]) ;
    xVector = zeros<vec>(vectorLevel*ElementData1.NumOfElementNodes[i]) ;
    CoordsTempStore = {join_rows(join_rows(xCoords, yCoords), zCoords)} ;
    for (Row=0; Row<ElementData1.NumOfElementNodes[i]; Row++)
    {
      Coords = arma::trans(CoordsTempStore(Row, span( 0, vectorLevel))) ;
      cosMatrix(Row, span((vectorLevel*Row-(vectorLevel-1)), (vectorLevel*Row))) = cos_xyz ;
      xVector(span((vectorLevel*Row-(vectorLevel-1)), (vectorLevel*Row)),0)= Coords ;
    }

}
#endif
